### PR TITLE
Linear interpolation (lerp) implementation

### PR DESCRIPTION
**Pull request by vertexclique**
Implements lerp with feature flag: #71015

### CODE REVIEW

**Comment by ebarnard** [comment]
The `(1 - t) * a + t * b` form of linear interpolation is only guarenteed to be monotonic if `a * b < 0`.

I've used the implementation from http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2018/p0811r2.html#linear-interpolation elsewhere, but this has the drawback that it is quite branch heavy.

**Comment by hanna-kruppe** [comment]
I find it hard to believe that the C++ version would see much use, given that it's much slower than the simpler alternatives and monotonicity does not seem useful for the most common uses of `lerp` that I'm familiar with.

I'd like to hear about applications where monotonicity matters, but regardless of whether a monotonic lerp is useful, I think it's valuable to have the fast and numerically stable `(1 - t) a + t b` readily available to non-experts. When people keep copying and pasting their own implementations for performance, some percentage of them will probably use the unstable `a + t (b - a)` instead.

I'm already worried that the proposed implementation might be eschewed by performance-minded users because it does not use FMA operations where available (which, to be fair, is currently difficult to do in libstd).

**Comment by ebarnard** [comment]
I agree that there are probably a limited number of stiuations that require strict monotinicity. However, there are undoubtedly some that do (I have seen a case where I *think* it mattered, and chaging to a monotonic implementation seemed to help fix the issue I was seeing).

Given that is the case, it seems odd for the standard library to include *this* implementation of `lerp` when there isn't a clear best implementation, and when choosing between implementations depends what the application is.

I would expect a game engine to use the `(1 - t) * a + t * b` form, while a mathematical library might use something more similar to the implementation I linked above.

**Comment by hanna-kruppe** [comment]
You could argue that the standard library should not include any `lerp` implementation, I don't feel strongly about that either way. But I think if it's going to be added at all, then the fast and reasonably-well-behaved variant is going to be more helpful to more people.

**Comment by XAMPPRocky** [comment]
> Given that is the case, it seems odd for the standard library to include this implementation of lerp when there isn't a clear best implementation, and when choosing between implementations depends what the application is.

The standard library often doesn't have the "best" implementation of a particular problem. There's usually always some trade-off. That being said, the guarantee should probably be mentioned in the documentation so that people are aware of the trade-offs being made.

**Comment by ebarnard** [comment]
Maybe my wording wasn't very good - by best I mean an implementation that is both correct and as fast as possible. Elsewhere, the standard library seems to favour correctness over speed (e.g. HashMap using SipHash by default). It would be odd for it not to do so here.

**Comment by ebarnard** [comment]
For reference [here](https://github.com/llvm-mirror/libcxx/blob/cae4caafdebfc30a35833e20f17e4d67668004b3/include/cmath#L616) is the implementation used in LLVM's libcxx. It is very similar to the one from [P0811R3].

Whatever implementation is decided on it ought to [pass the same tests](https://github.com/llvm-mirror/libcxx/blob/e84fd85b88bb1cc2c49179eeba6447c90829bf54/test/std/numerics/c.math/c.math.lerp/c.math.lerp.pass.cpp) as libcxx's.

[P0811R3]: http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2019/p0811r3.html

**Comment by hanna-kruppe** [comment]
Reality is a lot more complicated than a "correctness over speed" slogan suggests. Just one example thematically related to this PR: `Iterator::sum` with float elements performs naive serial summation, not anything more numerically robust such as compensated summation.

Furthermore, your arguments are based on a relatively arbitrary and contestable definition of correctness. For example, why does it matter if results are monotonic in `t` but it doesn't matter if results are correctly rounded? Why do the trade-offs w.r.t. numeric stability not factor into the conditions included in the C++ standard, other than the (literal, in the canonical implementation) special case of requiring `lerp(a, b, 1) == b`? There is a value judgement and trade-off underlying the decision to enshrine one property and disregard the others. Sometimes the right thing is to defer to experts or prior art on such decisions, but the C++ standard is not exactly an unimpeachable authority on numerics and std::lerp in particular has been highly controversial.

Finally, once again: if the end result is that libstd provides a function that most prospective users avoid over performance concerns, then little of value is gained. So in my view, your comments at most provide reasons to not provide `lerp` at all.

**Comment by crlf0710** [comment]
Let me reassign this to T-libs to make the decision here.
r? @Amanieu

**Comment by Amanieu** [comment]
I feel that if we do decide to include `lerp` in the standard library then we should use the "correct but slow" version of the function and have a separate `fast_lerp` for the fast but less precise version.

I have concerns regarding the argument order however: `a.lerp(b, t)` breaks the symmetry between the two values being interpolated over. I feel that `t.lerp(a, b)` would make much more sense. Alternatively we could also remove the method syntax and force all arguments to be specified: `lerp(a, b, t)`.

**Comment by ranma42** [comment]
Would it make sense to have `lerp` as a method on `Range<f32>` (or `RangeInclusive<f32>`)?

**Comment by vertexclique** [comment]
@cramertj Oh sorry I forgot to update you all. This is not ready for review yet, I need to refactor a couple of things. This version won't compile to the expected intrinsic in every rust project. So will use suboptimal non-fma ops. I will ping again. Thanks :)

**Comment by jdahlstrom** [comment]
It would make sense to me for `lerp` to mirror [`clamp`](https://github.com/rust-lang/rust/issues/44095), which is to say, `t.lerp(a, b)`. The asymmetry of `a.lerp(b, t)` does not feel right to me.

File: src/libstd/lib.rs [hunk 1]
```diff
@@ -247,6 +247,7 @@
 #![feature(cfg_target_thread_local)]
 #![feature(char_error_internals)]
 #![feature(clamp)]
+#![feature(lerp)]
 #![feature(concat_idents)]
```
Comment (dtolnay) [inline src/libstd/lib.rs:251]: This needs to be put in sorted order per the NB at the top and bottom of this feature list.
abcdefghijklmnopqrstuvwxyz

File: src/libstd/f32.rs [hunk 2]
```diff
@@ -912,6 +912,30 @@ impl f32 {
         }
         x
     }
+
+    /// Computes `self + step(upper - pol)` the linear interpolation between
```
Comment (dtolnay) [inline src/libstd/f32.rs:916]: Not sure what `step` refers to but this description doesn't look like it matches the implementation. In particular `pol` is [dimensionless](https://en.wikipedia.org/wiki/Dimensionless_quantity) and `upper` is not so `upper - pol` is not an expression that would make sense.

**Review by dtolnay — CHANGES_REQUESTED** [review]

**Comment by Dylan-DPC** [comment]
@vertexclique any updates?

**Comment by vertexclique** [comment]
Ok, I have no disk space left for compiling the compiler. I will carry on when I have a good machine at handy.

**Pull request closed without merging**